### PR TITLE
split ModelBase

### LINKED ISSUE #29: restructure ModelBase class
**Is your feature request related to a problem? Please describe.**
Restructure ModelBase class for better inheritance. 

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or task?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [x] 1. If the issue is a defect, is the defect fixed?
- [x] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [x] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [x] 4. If the issue is a defect, does it impact the latest release branch? If yes, is there any issue tagged with release (create if needed)?
- [x] 5. If the issue is being closed without a pull request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by wangcj05**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
close #29 

##### What are the significant changes in functionality due to this change request?

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual.
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Regression tests have to complete successfully.
- [x] 5. If significant functionality is added, there must be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test.
- [x] 6. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.

### CODE REVIEW

File: src/CapitalInvestments/PyomoModels/ModelBase.py [hunk 1]
```diff
@@ -216,6 +112,17 @@ def generateModelInputData(self):
     """
     pass
 
+  def processInputSets(self, indexName):
+    """
+      Method to generate Set input for pyomo model
+      @ In, indexName, str, name of index
+      @ Out, dict, {None:[indexValue]}
```
Comment (mandd) [inline src/CapitalInvestments/PyomoModels/ModelBase.py:119]: This is an unusual dict definition: why this structure?

File: src/CapitalInvestments/PyomoModels/ModelBase.py [hunk 2]
```diff
@@ -499,92 +305,6 @@ def printSolution(self, model):
     outputDict = {}
```
Comment (mandd) [inline src/CapitalInvestments/PyomoModels/ModelBase.py:305]: is this method needed given its function?

File: src/CapitalInvestments/PyomoModels/PySPBase.py [hunk 3]
```diff
@@ -0,0 +1,376 @@
...
+  @abc.abstractmethod
+  def pysp_scenario_tree_model_callback(self):
+    """
+      scenario tree instance creation callback
+      @ In, None
+      @ Out, treeModel, Instance, pyomo scenario tree model for two stage stochastic programming
+    """
+    pass
+
+  def pysp_instance_creation_callback(self, scenario_name, node_names):
```
Comment (mandd) [inline src/CapitalInvestments/PyomoModels/PySPBase.py:223]: method naming convention is not uniform throughout the class

**Review by mandd — CHANGES_REQUESTED** [review]
One final comment: could you provide a rationale of this main class restructuring?

File: src/CapitalInvestments/PyomoModels/ModelBase.py (hunk 1 above)
Comment (wangcj05) [reply src/CapitalInvestments/PyomoModels/ModelBase.py:119]: This is one format that pyomo accepts

File: src/CapitalInvestments/PyomoModels/ModelBase.py (hunk 2 above)
Comment (wangcj05) [reply src/CapitalInvestments/PyomoModels/ModelBase.py:305]: This function is overloaded by its inherited class.

File: src/CapitalInvestments/PyomoModels/PySPBase.py (hunk 3 above)
Comment (wangcj05) [reply src/CapitalInvestments/PyomoModels/PySPBase.py:223]: These are call back functions that are required by pyomo. We need to provide exactly the same name

**Review by mandd — APPROVED** [review]

**Pull request merged**